### DISCUSSION TITLE
Error when run sulcalpitsExtraction

### DISCUSSION THREAD

**Discussion by HusBunia** [brainvisa/brainvisa.github.io#139] [Q&A]
Dear BrainVisa expeerts : 

I'm still somewhat confused about some of the data formats, so my questions may seem a bit naive, please forgive me. I have run the sulcalpitextraction from axon-process. The command I used is 
`axon-runprocess SulcalPitsExtraction /casa/home/f01200624V005/surf/lh.white.gii left /casa/home/f01200624V005/surf/lh.white_pole_cingular.gii  0.03 1.5 20.0 236.18 50.0 91312.2 /casa/home/f01200624V005/surf/lh.white_DPF.gii /casa/home/f01200624V005/lh.white_pits.gii /casa/home/f01200624V005/surf/lh.white_noisy_pits.gii /casa/home/f01200624V005/surf/lh.white_ridges.gii /casa/home/f01200624V005/surf/lh.white_basins.gii`

lh.white.gii is converted by mris_convert from a freesurfer recon-all output.

But I fouod  it always return errors like below: 

`Initializing brainvisa... (takes a while)...

_Process **Sulcal Pits Extraction** started on 2023/05/19 23:07_

    Computing Vertex Voronoi
    -percent polygon with obtuse angle  20.59381214447579
computing DPF

    Computing Laplacian
    Computing mesh weights
    -edge length threshold needed for  0  values =  0.0  %
    -number of Nan in weights:  0  =  0.0  %
    -number of Negative values in weights:  32862  =  4.221585040620327  %
    -nb Nan in L :  0
    -nb Inf in L :  0
DPF done

Computing the Fiedler geodesic length and surface area

    Computing Laplacian
    Computing mesh weights
    -edge length threshold needed for  0  values =  0.0  %
    -number of Nan in weights:  0  =  0.0  %
    -number of Negative values in weights:  32862  =  4.221585040620327  %
    -nb Nan in L :  0
    -nb Inf in L :  0
Computing GEODESIC distance between the max and min
i: 778426number of the points in the path = 503, length of the path = 225.546

Computing the watershed with distance and ridge criteria for basins merging

i: 778426

ERROR: in _Sulcal Pits Extraction 1_ **
    TypeError: 'float' object cannot be interpreted as an integer
**
* * *
`**TypeError**`
    TypeError: 'float' object cannot be interpreted as an integer
---
_Process **Sulcal Pits Extraction** finished on 2023/05/19 23:09 (1 minutes 48
seconds)_

Traceback (most recent call last):
  File "/casa/install/bin/axon-runprocess", line 4, in <module>
    run_module('brainvisa.axon.runprocess')
  File "/usr/lib/python3.10/runpy.py", line 227, in run_module
    return _run_code(code, {}, init_globals, run_name, mod_spec)
  File "/usr/lib/python3.10/runpy.py", line 86, in _run_code
    exec(code, run_globals)
  File "/casa/install/python/brainvisa/axon/runprocess.py", line 512, in <module>
    run_process_with_distribution(
  File "/casa/install/python/brainvisa/axon/runprocess.py", line 270, in run_process_with_distribution
    brainvisa.processes.defaultContext().runProcess(process)
  File "/casa/install/python/brainvisa/processes.py", line 3120, in runProcess
    six.reraise(type(e[0]), e[0], e[1])
  File "/usr/lib/python3/dist-packages/six.py", line 719, in reraise
    raise value
  File "/casa/install/python/brainvisa/processes.py", line 3412, in _processExecution
    result = process.execution(self)
  File "/casa/install/brainvisa/toolboxes/cortical_surface/processes/anatomy/SulcalPitsExtraction.py", line 151, in execution
    labels_1, pitsKept_1, pitsRemoved_1, ridgePoints, parent_1 = watershed.watershed(mesh, vert_voronoi, dpf[0], mask, thresh_dist, self.thresh_ridge)
  File "/casa/install/python/brainvisa/cortical_surface/surface_tools/mesh_watershed.py", line 79, in watershed
    idx=np.arange((vert.size)/3).reshape((vert.size)/3,1) # vertices index
TypeError: 'float' object cannot be interpreted as an integer
Exception ignored in: <function TemporaryDiskItem.__del__ at 0x7f9f4558d900>
Traceback (most recent call last):
  File "/casa/install/python/brainvisa/data/neuroDiskItems.py", line 2756, in __del__
  File "/casa/install/python/soma/singleton.py", line 88, in __new__
TypeError: Application.__singleton_init__() missing 2 required positional arguments: 'name' and 'version'
Exception ignored in: <function TemporaryDirectory.__del__ at 0x7f9f4558d750>
Traceback (most recent call last):
  File "/casa/install/python/brainvisa/data/neuroDiskItems.py", line 2699, in __del__
TypeError: 'NoneType' object is not callable`

"I'm confused about which variable corresponds to the "NoneType" in the error messages, and what the 'float' object is. I'm using all default parameters, so I'm not sure if I've set some incorrect inputs?"

Really appreaciated for any help !

Thanks !

**Comment by denisri** [comment]
It seems that this code has not been properly ported and tested for python3. in the source file `/casa/install/python/brainvisa/cortical_surface/surface_tools/mesh_watershed.py` there are several occurrences of `vertex.size/3` or `(vertex.size)/3` which in python3 is a float result, and should be `vertex.shape[0]` instead.

**Comment by denisri** [comment]
I have pushed a fix in the sources but I don't know it it will be enough since I have not tested it and I have not run the process on my side.
The updated file can be found here:
https://brainvisa.info/download/updates/5.1.1/python/brainvisa/cortical_surface/surface_tools/mesh_watershed.py
To use it, you will have to perform a read/write install (I guess you are using a singularity/apptainer image of brainvisa):
run the `bv` command with no arguments from outside the container, it should open a GUI, where in the "Image" section you can perform a read/write install: it will copy all brainvisa files on the host filesystem, thus use about 2 GB of additional disk space. Then you can download the updated file above, and save/copy it at the same location in `<brainvisa_install_dir>/install/` (so here in `<brainvisa_install>/install/python/brainvisa/cortical_surface/surface_tools/`).
But as I said, there might be other problems later...

**Comment by HusBunia** [comment]
Hi densiri:

Thanks for your help ! !

I also tried run the python scripts in brainvisa-4.6.1 with python2.7, it gaves me the same error (seems the code was built in a more old version ). I found the easiest way to fix the problem is add .astype(int) after variables should be looped as an integer, a litter similar with your modified scripts. I guess the sulcal basins extraction pipeline in 4.6.1 and 5.1.1 seems not very different, is it ?  

after download files according to your instruction, I also seek a way to load brainvisa library in a python idel for convenient version error debug.  If I want to use the scripts in 5.1.1, I believe your worried would be true... I used bv command to load a idel like pycharm outsied the container, but only the soma-base were found in loaded enviroment. 

I greatly hope that a fixed version can be used in the near future ......

**Comment by denisri** [comment]
The sulcal basins extraction pipelines are not different between 4.6 and 5.1, you're right. There has been very little evolution in the toolboxes during the last years.
I'm a bit surprised the error was the same using python2.7, but anyway this code was kind of wrong, so it's better with the fix. Could you run it with the patched file (either mine or yours) ?
Using an IDE is possible, however there is the virtual container layer problem: the IDE itself runs outside of the container, but you need it to run code inside the container. Some IDEs can deal with this indirection, I know `VSCode` does it, I don't know for `PyCharm` (personally I do it the old way, using a source code editor which is not an IDE, and testing in an `ipython` shell, or using scripts, or a notebook, running inside the container). The problem is that code run from the IDE should not use the `python` command found in the host system path, but (at least) run `bv python`, or run an interactive environment inside the brainvisa container (`bv`).
Now if you are a developer and like to benefit from updates, you might consider installing a developer environment (using `casa-dev` containers and building from sources sync'd using git).
Last, for the `brainvisa` program (the `axon` project, actually), processes run from the GUI or via `axon-runprocess`, they need the axon environment loaded before the API can be used interactively. There are 2 ways of doing it;
- from the GUI, use the menu "File/Open shell": it will open an interactive `ipython` shell in a `qtconsole` widget
- from a regular shell (or `ipython`, or jupyter notebook): you need to run this code to initialize brainvisa/axon environment:
  ```
  from brainvisa.axon import processes
  processes.initializeProcesses()
  from brainvisa.processes import *
  ```
  then you can instantiate and run brainvisa processes. I don't know if it's what you were asking about...

**Reply by HusBunia** [reply]
In version 4.6.1, after adding the necessary `.astype(int),` the code runs successfully. After executing the watershed and area filtering with default parameters, the resulting single subject basin atlas looks like this:
<img width="578" alt="a4d3f82394522e16c36095c3851b0b7" src="https://github.com/brainvisa/brainvisa.github.io/assets/40214939/dcb5eda7-a9bf-4162-8f75-bea15cd91ccc">
I haven't run a complete watershed in Python 3 yet (as it's quite time-consuming..). It seems that the adaptive watershed algorithm could also provide me with the correct group brain atlas because it calls the same function just replace the `vert_area` with basin density map. 

Thank you for your suggestion, I learned a lot. VSCode is more suitable for my current machine, at least for my current task. 

Thanks for your help again !
